Funqy QueryObjectReader not working in native image.

**Describe the bug**

Funqy QueryObjectReader not working in native image.

Note that jackson object reader works fine.

**Expected behavior**

```shell
curl -v "http://localhost:8080/greet?name=Matej" 
*   Trying ::1:8080...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> GET /greet?name=Matej HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.66.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Content-Type: application/json
< content-length: 41
< 
* Connection #0 to host localhost left intact
{"name":"Matej","message":"Hello Matej!"}
```

**Actual behavior**
To reproduce call sample greeting function:
```shell
curl -v "http://localhost:8080/greet?name=Matej"              
*   Trying ::1:8080...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> GET /greet?name=Matej HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.66.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 400 Bad Request
< content-length: 0
< 
* Connection #0 to host localhost left intact

```

In the application log:
```
2020-09-10 01:00:37,280 INFO  [io.quarkus] (main) funq-demo 1.0-SNAPSHOT native (powered by Quarkus 1.8.0.CR1) started in 0.029s. Listening on: http://0.0.0.0:8080
2020-09-10 01:00:37,280 INFO  [io.quarkus] (main) Profile prod activated. 
2020-09-10 01:00:37,280 INFO  [io.quarkus] (main) Installed features: [cdi, funqy-http]
2020-09-10 01:00:39,752 ERROR [io.qua.funqy] (vert.x-eventloop-thread-0) Failed to unmarshal input: java.lang.RuntimeException: java.lang.InstantiationException: Type `org.funqy.demo.Identity` can not be instantiated reflectively as it does not have a no-parameter constructor or the no-parameter constructor has not been added explicitly to the native image.                                                 











        at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:133)
        at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:54)
        at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:36)
        at io.quarkus.vertx.http.runtime.VertxHttpRecorder$9.handle(VertxHttpRecorder.java:392)
        at io.quarkus.vertx.http.runtime.VertxHttpRecorder$9.handle(VertxHttpRecorder.java:389)
        at io.quarkus.vertx.http.runtime.VertxHttpRecorder$1.handle(VertxHttpRecorder.java:133)
        at io.quarkus.vertx.http.runtime.VertxHttpRecorder$1.handle(VertxHttpRecorder.java:122)
        at io.vertx.core.http.impl.WebSocketRequestHandler.handle(WebSocketRequestHandler.java:50)
        at io.vertx.core.http.impl.WebSocketRequestHandler.handle(WebSocketRequestHandler.java:32)
        at io.vertx.core.http.impl.Http1xServerConnection.handleMessage(Http1xServerConnection.java:136)
        at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:366)
        at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:43)
        at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:229)
        at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:173)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
        at io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.channelRead(WebSocketServerExtensionHandler.java:101)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.vertx.core.http.impl.Http1xUpgradeToH2CHandler.channelRead(Http1xUpgradeToH2CHandler.java:109)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.vertx.core.http.impl.Http1xOrH2CHandler.end(Http1xOrH2CHandler.java:61)
        at io.vertx.core.http.impl.Http1xOrH2CHandler.channelRead(Http1xOrH2CHandler.java:38)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:834)
        at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:517)
        at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:193)
Caused by: java.lang.InstantiationException: Type `org.funqy.demo.Identity` can not be instantiated reflectively as it does not have a no-parameter constructor or the no-parameter constructor has not been added explicitly to the native image.
        at java.lang.Class.newInstance(DynamicHub.java:830)
        at io.quarkus.funqy.runtime.query.QueryObjectReader.create(QueryObjectReader.java:22)
        ... 69 more
```

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/ReflectiveHierarchyStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/ReflectiveHierarchyStep.java
@@ -165,6 +165,9 @@ public class ReflectiveHierarchyStep {
             Map<DotName, Set<String>> unindexedClasses,
             Predicate<ClassInfo> finalFieldsWritable,
             BuildProducer<ReflectiveClassBuildItem> reflectiveClass) {
+        if (name == null) {
+            return;
+        }
         if (reflectiveHierarchyBuildItem.getIgnoreTypePredicate().test(name)) {
             return;
         }

--- a/extensions/funqy/funqy-server-common/deployment/src/main/java/io/quarkus/funqy/deployment/FunctionScannerBuildStep.java
+++ b/extensions/funqy/funqy-server-common/deployment/src/main/java/io/quarkus/funqy/deployment/FunctionScannerBuildStep.java
@@ -1,6 +1,7 @@
 package io.quarkus.funqy.deployment;
 
 import static io.quarkus.deployment.annotations.ExecutionTime.STATIC_INIT;
+import static io.quarkus.funqy.deployment.ReflectionRegistrationUtil.*;
 
 import java.lang.reflect.Modifier;
 import java.util.Collection;
@@ -14,6 +15,7 @@ import org.jboss.jandex.AnnotationTarget;
 import org.jboss.jandex.AnnotationValue;
 import org.jboss.jandex.ClassInfo;
 import org.jboss.jandex.DotName;
+import org.jboss.jandex.IndexView;
 import org.jboss.jandex.MethodInfo;
 import org.jboss.jandex.Type;
 import org.objectweb.asm.ClassVisitor;
@@ -32,13 +34,16 @@ import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.annotations.Record;
 import io.quarkus.deployment.builditem.BytecodeTransformerBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
+import io.quarkus.deployment.builditem.nativeimage.ReflectiveHierarchyBuildItem;
 import io.quarkus.deployment.recording.RecorderContext;
+import io.quarkus.funqy.Context;
 import io.quarkus.funqy.Funq;
 import io.quarkus.funqy.runtime.FunctionRecorder;
 import io.quarkus.gizmo.Gizmo;
 
 public class FunctionScannerBuildStep {
     public static final DotName FUNQ = DotName.createSimple(Funq.class.getName());
+    public static final DotName CONTEXT = DotName.createSimple(Context.class.getName());
 
     @BuildStep
     public void scanFunctions(BeanArchiveIndexBuildItem beanArchiveIndexBuildItem,
@@ -46,8 +51,10 @@ public class FunctionScannerBuildStep {
             BuildProducer<AnnotationsTransformerBuildItem> annotationsTransformer,
             BuildProducer<UnremovableBeanBuildItem> unremovableBeans,
             BuildProducer<ReflectiveClassBuildItem> reflectiveClass,
+            BuildProducer<ReflectiveHierarchyBuildItem> reflectiveHierarchy,
             BuildProducer<FunctionBuildItem> functions) {
-        Collection<AnnotationInstance> funqs = beanArchiveIndexBuildItem.getIndex().getAnnotations(FUNQ);
+        IndexView index = beanArchiveIndexBuildItem.getIndex();
+        Collection<AnnotationInstance> funqs = index.getAnnotations(FUNQ);
         Set<ClassInfo> classes = new HashSet<>();
         Set<String> classNames = new HashSet<>();
         for (AnnotationInstance funqMethod : funqs) {
@@ -63,6 +70,34 @@ public class FunctionScannerBuildStep {
             if (functionName != null && functionName.isEmpty())
                 functionName = null;
             functions.produce(new FunctionBuildItem(className, methodName, functionName));
+
+            String source = FunctionScannerBuildStep.class.getSimpleName() + " > " + method.declaringClass() + "[" + method
+                    + "]";
+
+            Type returnType = method.returnType();
+            if (returnType.kind() != Type.Kind.VOID) {
+                reflectiveHierarchy.produce(new ReflectiveHierarchyBuildItem.Builder()
+                        .type(returnType)
+                        .index(index)
+                        .ignoreTypePredicate(IGNORE_TYPE_FOR_REFLECTION_PREDICATE)
+                        .ignoreFieldPredicate(IGNORE_FIELD_FOR_REFLECTION_PREDICATE)
+                        .ignoreMethodPredicate(IGNORE_METHOD_FOR_REFLECTION_PREDICATE)
+                        .source(source)
+                        .build());
+            }
+            for (short i = 0; i < method.parameters().size(); i++) {
+                Type parameterType = method.parameters().get(i);
+                if (!hasAnnotation(method, i, CONTEXT)) {
+                    reflectiveHierarchy.produce(new ReflectiveHierarchyBuildItem.Builder()
+                            .type(parameterType)
+                            .index(index)
+                            .ignoreTypePredicate(IGNORE_TYPE_FOR_REFLECTION_PREDICATE)
+                            .ignoreFieldPredicate(IGNORE_FIELD_FOR_REFLECTION_PREDICATE)
+                            .ignoreMethodPredicate(IGNORE_METHOD_FOR_REFLECTION_PREDICATE)
+                            .source(source)
+                            .build());
+                }
+            }
         }
         Set<ClassInfo> withoutDefaultCtor = new HashSet<>();
         for (ClassInfo clazz : classes) {
@@ -94,6 +129,18 @@ public class FunctionScannerBuildStep {
                 transformation.done();
             }
         }));
+    }
+
+    private static boolean hasAnnotation(MethodInfo method, short paramPosition, DotName annotation) {
+        for (AnnotationInstance annotationInstance : method.annotations()) {
+            AnnotationTarget target = annotationInstance.target();
+            if (target != null && target.kind() == AnnotationTarget.Kind.METHOD_PARAMETER
+                    && target.asMethodParameter().position() == paramPosition
+                    && annotationInstance.name().equals(annotation)) {
+                return true;
+            }
+        }
+        return false;
     }
 
     @BuildStep

--- a/extensions/funqy/funqy-server-common/deployment/src/main/java/io/quarkus/funqy/deployment/ReflectionRegistrationUtil.java
+++ b/extensions/funqy/funqy-server-common/deployment/src/main/java/io/quarkus/funqy/deployment/ReflectionRegistrationUtil.java
@@ -1,0 +1,63 @@
+package io.quarkus.funqy.deployment;
+
+import java.util.function.Predicate;
+
+import org.jboss.jandex.DotName;
+import org.jboss.jandex.FieldInfo;
+import org.jboss.jandex.MethodInfo;
+
+import io.quarkus.deployment.builditem.nativeimage.ReflectiveHierarchyBuildItem;
+
+final class ReflectionRegistrationUtil {
+
+    private ReflectionRegistrationUtil() {
+    }
+
+    static final IgnoreTypeForReflectionPredicate IGNORE_TYPE_FOR_REFLECTION_PREDICATE = new IgnoreTypeForReflectionPredicate();
+    static final IgnoreFieldForReflectionPredicate IGNORE_FIELD_FOR_REFLECTION_PREDICATE = new IgnoreFieldForReflectionPredicate();
+    static final IgnoreMethodForReflectionPredicate IGNORE_METHOD_FOR_REFLECTION_PREDICATE = new IgnoreMethodForReflectionPredicate();
+
+    private static final DotName JSON_IGNORE = DotName.createSimple("com.fasterxml.jackson.annotation.JsonIgnore");
+    private static final DotName JSONB_TRANSIENT = DotName.createSimple("javax.json.bind.annotation.JsonbTransient");
+
+    private static class IgnoreTypeForReflectionPredicate implements Predicate<DotName> {
+
+        @Override
+        public boolean test(DotName dotName) {
+            if (ReflectiveHierarchyBuildItem.DefaultIgnoreTypePredicate.INSTANCE.test(dotName)) {
+                return true;
+            }
+
+            String name = dotName.toString();
+            for (String packageName : PACKAGES_IGNORED_FOR_REFLECTION) {
+                if (name.startsWith(packageName)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+    }
+
+    private static class IgnoreFieldForReflectionPredicate implements Predicate<FieldInfo> {
+
+        @Override
+        public boolean test(FieldInfo fieldInfo) {
+            return fieldInfo.hasAnnotation(JSON_IGNORE)
+                    || fieldInfo.hasAnnotation(JSONB_TRANSIENT);
+        }
+    }
+
+    private static class IgnoreMethodForReflectionPredicate implements Predicate<MethodInfo> {
+
+        @Override
+        public boolean test(MethodInfo methodInfo) {
+            return methodInfo.hasAnnotation(JSON_IGNORE)
+                    || methodInfo.hasAnnotation(JSONB_TRANSIENT);
+        }
+    }
+
+    private static final String[] PACKAGES_IGNORED_FOR_REFLECTION = {
+            // Mutiny
+            "io.smallrye.mutiny."
+    };
+}

--- a/extensions/resteasy-common/spi/src/main/java/io/quarkus/resteasy/common/spi/ResteasyDotNames.java
+++ b/extensions/resteasy-common/spi/src/main/java/io/quarkus/resteasy/common/spi/ResteasyDotNames.java
@@ -112,6 +112,8 @@ public final class ResteasyDotNames {
             // RESTEasy
             "org.jboss.resteasy.",
             // Vert.x JSON layer
-            "io.vertx.core.json."
+            "io.vertx.core.json.",
+            // Mutiny
+            "io.smallrye.mutiny."
     };
 }

